fem/libceed/ceed-wrappers.cpp

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "ceed-wrappers.hpp"
+
+#ifdef MFEM_USE_CEED
+#include <ceed-backend.h>
+#include "ceed.hpp"
+#include "ceedsolvers-utility.h"
+
+namespace mfem
+{
+
+class UnconstrainedMFEMCeedOperator : public Operator
+{
+public:
+   UnconstrainedMFEMCeedOperator(CeedOperator oper);
+   ~UnconstrainedMFEMCeedOperator();
+   virtual void Mult(const Vector& x, Vector& y) const;
+   CeedOperator GetCeedOperator() const { return oper_; }
+   using Operator::SetupRAP;
+private:
+   CeedOperator oper_;
+   CeedVector u_, v_;
+};
+
+UnconstrainedMFEMCeedOperator::UnconstrainedMFEMCeedOperator(CeedOperator oper)
+   : oper_(oper)
+{
+   int ierr = 0;
+   Ceed ceed;
+   ierr += CeedOperatorGetCeed(oper, &ceed);
+   CeedElemRestriction er;
+   ierr += CeedOperatorGetActiveElemRestriction(oper, &er);
+   int s;
+   ierr += CeedElemRestrictionGetLVectorSize(er, &s);
+   height = width = s;
+   ierr += CeedVectorCreate(ceed, height, &v_);
+   ierr += CeedVectorCreate(ceed, width, &u_);
+   MFEM_ASSERT(ierr == 0, "CEED error");
+}
+
+UnconstrainedMFEMCeedOperator::~UnconstrainedMFEMCeedOperator()
+{
+   int ierr = 0;
+   ierr += CeedVectorDestroy(&v_);
+   ierr += CeedVectorDestroy(&u_);
+   MFEM_ASSERT(ierr == 0, "CEED error");
+}
+
+void UnconstrainedMFEMCeedOperator::Mult(const Vector& x, Vector& y) const
+{
+   int ierr = 0;
+
+   const CeedScalar *x_ptr;
+   CeedScalar *y_ptr;
+   CeedMemType mem;
+   CeedGetPreferredMemType(internal::ceed, &mem);
+   if (Device::Allows(Backend::CUDA) && mem == CEED_MEM_DEVICE)
+   {
+      x_ptr = x.Read();
+      y_ptr = y.ReadWrite();
+   }
+   else
+   {
+      x_ptr = x.HostRead();
+      y_ptr = y.HostReadWrite();
+      mem = CEED_MEM_HOST;
+   }
+
+   ierr += CeedVectorSetArray(u_, mem, CEED_USE_POINTER,
+                              const_cast<CeedScalar*>(x_ptr));
+   ierr += CeedVectorSetArray(v_, mem, CEED_USE_POINTER, y_ptr);
+
+   ierr += CeedOperatorApply(oper_, u_, v_, CEED_REQUEST_IMMEDIATE);
+
+   ierr += CeedVectorTakeArray(u_, mem, const_cast<CeedScalar**>(&x_ptr));
+   ierr += CeedVectorTakeArray(v_, mem, &y_ptr);
```

Comment:
It would be nice to just use `CeedAddMult`

Reply:
Fixed, but my solution is a little bit of a hack. Let me know if you have a better idea.

Reply:
I made some more changes and think it's better.